[quant] Adding zero point type check for per channel quantization

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#40811 [quant] Adding zero point type check for per channel quantization**

Differential Revision: [D22319417](https://our.internmc.facebook.com/intern/diff/D22319417)

Commit:
[quant] Adding zero point type check for per channel quantization

[ghstack-poisoned]

Commit:
Update on "[quant] Adding zero point type check for per channel quantization"


Differential Revision: [D22319417](https://our.internmc.facebook.com/intern/diff/D22319417)

[ghstack-poisoned]

Commit:
Update on "[quant] Adding zero point type check for per channel quantization"


Differential Revision: [D22319417](https://our.internmc.facebook.com/intern/diff/D22319417)

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/native/quantized/fake_quant_per_channel_affine.cpp:
```diff
@@ -35,6 +35,10 @@ Tensor fake_quantize_per_channel_affine(
     int64_t quant_min,
     int64_t quant_max) {
   TORCH_CHECK(self.scalar_type() == ScalarType::Float);
+  TORCH_CHECK(scale.scalar_type() == ScalarType::Float,
```
double I think?

z-a-f replied:
I think it's float, as per the iterator expectation [here](https://github.com/pytorch/pytorch/blob/4cf81b32cb1aacb543980ae751b99aca22b3fc31/aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp#L2037). Should I change them to double?

jerryzh168 replied:
Oh I thought this is quantize_per_tensor.

If this is float then probably zero_point should be int32_t, just to be consistent.

z-a-f replied:
The problem is that in the kernel implementation it is `float` for scale and `int64_t` for the zero_point. I will have to change all the kernels

jerryzh168 replied:
do you know why is that? it would be better to be consistent

z-a-f replied:
I am not sure why the inconsistency, but I see some kernels use 64 bits and some 32.

Commit:
Update on "[quant] Adding zero point type check for per channel quantization"


Differential Revision: [D22319417](https://our.internmc.facebook.com/intern/diff/D22319417)

[ghstack-poisoned]

Commit:
Update on "[quant] Adding zero point type check for per channel quantization"


Differential Revision: [D22319417](https://our.internmc.facebook.com/intern/diff/D22319417)

[ghstack-poisoned]

jerryzh168 reviewed (approved):
Please make sure to change int64_t to int32_t in follow up PRs

Commit:
Update on "[quant] Adding zero point type check for per channel quantization"


Differential Revision: [D22319417](https://our.internmc.facebook.com/intern/diff/D22319417)

[ghstack-poisoned]

Commit:
Update on "[quant] Adding zero point type check for per channel quantization"


Differential Revision: [D22319417](https://our.internmc.facebook.com/intern/diff/D22319417)

[ghstack-poisoned]
